app/models/invitation_token.rb

```diff
@@ -50,6 +53,10 @@ def token=(new_token)
 
   def expired?
     # If a response is still valid, it should always be possible to fill it out.
-    response.response_expired? && Time.zone.now > expires_at
+    return false if Time.zone.now <= expires_at
+    invitation_set.responses.each do |response|
+      return false unless response.response_expired?
```

Comment:
is deze check eigenlijk nog nodig? aangezien je de `expired_at` update aan de hand van de `expires_at` van het response

Reply:
Das een hele scherpe opmerking. Ik had m erin gelaten maar mijn redenering klopt niet.

Het verschil is namelijk dat we hier ook completed responses meetellen, een bij de standaard expires_at check van een invitation_token tellen we completed responses niet mee bij het berekenen van de expires_at. Deze check zegt dus als het ware dat je op het token van een ingevulde vragenlijst nog kunt inloggen tot aan het moment waarop die vragenlijst eigenlijk zou expiren. Maar dat is natuurlijk niet zo, omdat we iedere nacht de geexpirede invitation_tokens gewoon destroyen, en bij die expired check tellen we completed responses dus niet mee. Dus dit scenario wat ik hierboven beschrijf (dat je meer dan een week na het invullen van een vragenlijst die bijvoorbeeld de hele studie open blijft nog kan inloggen) zal nooit voorkomen omdat dan het bijbehorende invitation_token dan al eerder is opgeruimd.

We moeten dus een keuze maken: willen we invitation_tokens die horen bij gecomplete responses pas opruimen als de oorspronkelijke measurement_duration verstreken is? Dus voor metingen zoals een voormeting (die een measurement_duration van nil hebben wat betekent dat ze pas openhouden aan het einde van het protocol), betekent dat dat dat invitation_token van de voormeting nog maandenlang geldig blijft, ookal heb je de vragenlijst ingevuld. Misschien willen we dit voor mentoren dat ze hun vragenlijsten nog kunnen editen in de toegestane tijd oid? Of zeggen we: zodra je iets invult heb je nog 7 dagen om het te editten (dwz nog 7 dagen dat de invitation_token geldig blijft) en daarna gooien we het invitation_token weg. Ik neig naar het laatste, omdat dat simpeler is en je dan niet met invitation_tokens zit die superlang geldig zijn voor vragenlijsten die al zijn ingevuld (zolang een vragenlijst niet is ingevuld kan het invitation_token in beide scenarios wel gewoon maandenlang bestaan).

Mij lijkt de tweede optie het best.
 
Als we gaan voor de eerste optie moet ik de logica van de calculate_expires_at functie aanpassen, en in het tweede geval moet ik de logica van de expired? functie aanpassen.